### ISSUE TITLE
Collocating optional list and zipper list in the same package

### ISSUE THREAD

**Issue by gyzerok** [NoRedInk/list-selection#8]
Hi folks!

I am creating this issue out from small discussion on slack with @rtfeldman.

I’ve just started using `list-selection` and I love it! Also I knew that Richard had module similar to it but with additional constraint of item always being selected. I’ve checked Richards package and found one thing that might worth improving.

Currently `list-selection` exposes `List.Selection` module, while Richards `selectlist` exposes `SelectList` module. These 2 packages seem to be companions. They have similar idea, just a bit different set of constraints. If one happen to use them in the same project it’ll be a bit tricky especially for new people without Elm experience. Also it seems to be in the spirit of Elm to try to provide more recognizable naming.

My proposal would be to extend this package with zipper list. It will be similar to `elm-all-dict` from Noah which exposes `AllDict` and `EveryDict`.

**Comment by gyzerok** [comment]
Some additional benefit is It'll be easier to ensure both lists have similarish API. For example if some function makes sense for both it could be done as a single PR.

For example it got discovered that there is need for `map` function. It kinda make sense to have it for both packages. Then one can make API changes for both of them in single release.

I believe that samish API in this case it beneficial because it helps building intuition. By knowing one list folks will know another. And then decision comes more towards constraints they want to embrace by using one or another.

**Comment by BrianHicks** [comment]
I think we're going to have to use a zipper under the covers sooner or later, as we don't handle duplicate items very well.

**Comment by gyzerok** [comment]
Is there an update on this issue? I was thinking that it might be not a good idea to create one more library with suggestions I’ve provided. However if there are no plans to update this library, I will probably go with a fork.

**Comment by BrianHicks** [comment]
Hmm, no I just hadn't gotten back to it. @rtfeldman what do you think? Move to yours or move your stuff here?

**Comment by rtfeldman** [comment]
I went through both APIs, and I don't think it makes sense to merge these after all. 

In one data structure, the whole point is the guarantee that exactly 1 item is selected. In the other, the whole point is the guarantee that either 0 or 1 are selected.

One irreconcilable API difference resulting from this is that both have a function called `selected`, but they have different types. [One](http://package.elm-lang.org/packages/NoRedInk/list-selection/1.3.0/List-Selection#selected) has the type `DataStructure a -> Maybe a`, wheres [the other](http://package.elm-lang.org/packages/rtfeldman/selectlist/1.0.0/SelectList#selected) has the type `DataStructure a -> a`.

Because of this, it also doesn't really make sense for them to share a common implementation under the hood. One wants to look like this:

```elm
type DataStructure a
    = DataStructure (List a) a (List a)
```

...whereas the other wants, at a minimum, a `Maybe a` in that middle slot that's currently an `a`.

My conclusion is that these packages are as different as `List` and `NonEmptyList`, and should be separate packages.

That said, it might be wise coordinate the packages' names? It would be nice if you could guess the difference between them just from their names... 🤔

**Comment by BrianHicks** [comment]
something like `List.Selection` / `List.Selection.Optional`?

**Comment by gyzerok** [comment]
> My conclusion is that these packages are as different as `List` and `NonEmptyList`, and should be separate packages.

Totally agree with this comparison.

However I am not sure if `List` and `NonEmptyList` are in separate packages because of this difference. To me it seems like they are in the separate packages just because one of them is part of the `core` and there is simply no other way.

Another possible example I can give - [`elm-time`](http://package.elm-lang.org/packages/elm-community/elm-time). It contains both `Date` and `DateTime`. Their API is not exactly the same, but they represent very similar concepts. In my opinion one can say the same about `List`/`NonEmptyList` and data structures we are discussing here.

In other words I am not convinced they should be in separate packages 😄  But after all, decision is not mine to make. If you are not convinced, coordinating their names would be nice.
